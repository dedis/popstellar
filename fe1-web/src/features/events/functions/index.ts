export * from './events';
export * from './ui';
