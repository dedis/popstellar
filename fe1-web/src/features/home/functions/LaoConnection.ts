import { Hash } from 'core/objects';

import { ConnectToLao } from '../objects';

/**
 * Given a list of lao server addresses and the lao id, this computes the data
 * that is encoded in a QR code that can be used to connect to a LAO
 * @returns Encoded LAO connection
 */
export const encodeLaoConnectionForQRCode = (servers: string[], laoId: Hash): string =>
  new ConnectToLao({ servers, lao: laoId }).toJson();
