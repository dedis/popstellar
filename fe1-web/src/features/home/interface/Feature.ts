export namespace HomeFeature {
  export interface LaoState {
    id: string;
  }

  export interface Lao {
    toState: () => LaoState;
  }

  export interface Screen {
    id: string;
    title?: string;
    Component: React.ComponentType<unknown>;
    order: number;
  }
}
