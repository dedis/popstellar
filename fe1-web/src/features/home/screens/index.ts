export { default as Home } from './Home';
export { default as Launch } from './Launch';
export { default as ConnectConfirm } from './ConnectConfirm';
export { default as ConnectScan } from './ConnectScan';
