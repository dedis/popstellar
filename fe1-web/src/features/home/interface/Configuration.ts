import React from 'react';
import { AnyAction } from 'redux';

import { Channel, Hash } from 'core/objects';
import FeatureInterface from 'core/objects/FeatureInterface';

import { HomeFeature } from './Feature';

export const HOME_FEATURE_IDENTIFIER = 'home';

export interface HomeCompositionConfiguration {
  /* lao */

  /* functions */
  requestCreateLao: (laoName: string) => Promise<Channel>;
  connectToTestLao: () => void;

  /* action creators */
  addLaoServerAddress: (laoId: Hash, address: string) => AnyAction;

  /* hooks */

  /**
   * Gets the current lao list
   * @returns The current lao list
   */
  useLaoList: () => HomeFeature.Lao[];

  /* components */

  /**
   * A component rendering the list of recent laos
   */
  LaoList: React.ComponentType<unknown>;

  /* other */

  /**
   * A list of screens show in the main navigations
   */
  mainNavigationScreens: HomeFeature.Screen[];
}

/**
 * The type of the context that is provided to react home components
 */
export type HomeReactContext = Pick<
  HomeCompositionConfiguration,
  /* lao */
  | 'requestCreateLao'
  | 'addLaoServerAddress'
  | 'connectToTestLao'
  | 'useLaoList'
  | 'LaoList'
  | 'mainNavigationScreens'
>;

/**
 * The interface the home feature exposes
 */
export interface HomeInterface extends FeatureInterface {
  navigation: {
    MainNavigation: React.ComponentType<any>;
  };
  screens: {
    Home: React.ComponentType<any>;
    Launch: React.ComponentType<any>;
  };
  context: HomeReactContext;
}
