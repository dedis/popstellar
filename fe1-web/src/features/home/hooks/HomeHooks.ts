import { useContext } from 'react';

import FeatureContext from 'core/contexts/FeatureContext';

import { HomeReactContext, HOME_FEATURE_IDENTIFIER } from '../interface';

export namespace HomeHooks {
  export const useHomeContext = (): HomeReactContext => {
    const featureContext = useContext(FeatureContext);
    // assert that the evoting context exists
    if (!(HOME_FEATURE_IDENTIFIER in featureContext)) {
      throw new Error('Home context could not be found!');
    }
    return featureContext[HOME_FEATURE_IDENTIFIER] as HomeReactContext;
  };

  /**
   * Gets the action creator to connect to a lao
   * @returns The action creator
   */
  export const useConnectToTestLao = () => useHomeContext().connectToTestLao;

  /**
   * Gets the function that sends a network request to create a new lao
   * @returns The function to create a lao
   */
  export const useRequestCreateLao = () => useHomeContext().requestCreateLao;

  /**
   * Gets the function to add a server address to a lao
   * @returns The function to add a server address
   */
  export const useAddLaoServerAddress = () => useHomeContext().addLaoServerAddress;

  /**
   * Gets the current lao list
   * @returns The current lao list
   */
  export const useLaoList = () => useHomeContext().useLaoList();

  /**
   * Gets the lao list component
   * @returns The lao list component
   */
  export const useLaoListComponent = () => useHomeContext().LaoList;

  /**
   * Gets the list of screens to be rendered in the main navigation
   * @returns The list of screens
   */
  export const useHomeNavigationScreens = () => useHomeContext().homeNavigationScreens;

  /**
   * Returns a function from the context for obtaining the channel for a given lao
   * @returns A function for getting the channel by lao id
   */
  export const useGetLaoChannel = () => useHomeContext().getLaoChannel;

  /**
   * A hook returning if currently connected to a lao
   * Returns undefined if there is no current lao
   */
  export const useConnectedToLao = () => useHomeContext().useConnectedToLao();

  /**
   * Gets the disconnect from lao function
   */
  export const useDisonnectFromLao = () => useHomeContext().useDisconnectFromLao();

  /**
   * Gets the function to retrieve a lao by its id
   */
  export const useGetLaoById = () => useHomeContext().getLaoById;

  /**
   * Gets the function to resubscribe to a lao
   */
  export const useResubscribeToLao = () => useHomeContext().resubscribeToLao;
}
