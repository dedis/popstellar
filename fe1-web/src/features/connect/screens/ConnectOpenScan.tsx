import { useNavigation } from '@react-navigation/core';
import React, { useEffect, useState } from 'react';
import { View } from 'react-native';
import { useToast } from 'react-native-toast-notifications';
import QrReader from 'react-qr-reader';

import { WideButtonView } from 'core/components';
import containerStyles from 'core/styles/stylesheets/containerStyles';
import { FOUR_SECONDS } from 'resources/const';
import STRINGS from 'resources/strings';

import { ConnectHooks } from '../hooks';
import { ConnectToLao } from '../objects';

/**
 * Starts a QR code scan
 */
const ConnectOpenScan = () => {
  // FIXME: route should use proper type
  const navigation = useNavigation<any>();

  // Remove the user to go back to the ConnectEnableCamera as he has already given
  // his permission to use the camera

  // this is needed as otherwise the camera will stay turned on
  const [showScanner, setShowScanner] = useState(false);
  // re-enable scanner on focus events
  useEffect(() => {
    // Return the function to unsubscribe from the event so it gets removed on unmount
    return navigation.addListener('focus', () => {
      // The screen is focused, set QrWasScanned to false (i.e. allow scanner to be reused)
      setShowScanner(true);
    });
  }, [navigation]);

  const toast = useToast();

  const laoId = ConnectHooks.useCurrentLaoId();

  const handleError = (err: string) => {
    console.error(err);
    toast.show(err, {
      type: 'danger',
      placement: 'top',
      duration: FOUR_SECONDS,
    });
  };

  const handleScan = (data: string | null) => {
    if (!data) {
      return;
    }

    try {
      const obj = JSON.parse(data);
      const connectToLao = ConnectToLao.fromJson(obj);

      // if we are already connected to a LAO, then only allow new connections
      // to servers for the same LAO id
      if (laoId && connectToLao.lao !== laoId.valueOf()) {
        toast.show(
          `The scanned QR code is for a different LAO than the one currently connected to`,
          {
            type: 'warning',
            placement: 'top',
            duration: FOUR_SECONDS,
          },
        );
        return;
      }

      setShowScanner(false);
      navigation.navigate(STRINGS.connect_confirm_title, {
        laoIdIn: connectToLao.lao,
        url: connectToLao.server,
      });
    } catch (error) {
      toast.show(STRINGS.connect_scanning_fail, {
        type: 'danger',
        placement: 'top',
        duration: FOUR_SECONDS,
      });
    }
  };

  const onCancel = () => {
    setShowScanner(false);

    // if we have an active lao, this was an additional connection and thus we navigate (back)
    // to the organization user screen
    if (laoId) {
      navigation.navigate(STRINGS.app_navigation_tab_user, {
        screen: STRINGS.organization_navigation_tab_user,
      });
    } else {
      navigation.navigate(STRINGS.connect_unapproved_title);
    }
  };

  return showScanner ? (
    <View style={containerStyles.centeredXY}>
      <QrReader delay={300} onError={handleError} onScan={handleScan} style={{ width: '30%' }} />
      <WideButtonView title={STRINGS.general_button_cancel} onPress={onCancel} />
    </View>
  ) : (
    <View style={containerStyles.centeredY} />
  );
};

export default ConnectOpenScan;
