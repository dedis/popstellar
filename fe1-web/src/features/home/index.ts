import { HomeCompositionConfiguration, HomeInterface, HOME_FEATURE_IDENTIFIER } from './interface';
import * as navigation from './navigation';
import * as screens from './screens';

/**
 * Configures the Home feature
 */
export function compose(config: HomeCompositionConfiguration): HomeInterface {
  return {
    identifier: HOME_FEATURE_IDENTIFIER,
    navigation,
    screens,
    context: {
      requestCreateLao: config.requestCreateLao,
      addLaoServerAddress: config.addLaoServerAddress,
      connectToTestLao: config.connectToTestLao,
      useLaoList: config.useLaoList,
      LaoList: config.LaoList,
      mainNavigationScreens: config.mainNavigationScreens,
    },
  };
}
