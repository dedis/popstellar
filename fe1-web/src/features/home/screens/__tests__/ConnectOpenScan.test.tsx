import { render } from '@testing-library/react-native';
import React from 'react';
// @ts-ignore
import { Provider } from 'react-redux';
import { combineReducers, createStore } from 'redux';

import MockNavigator from '__tests__/components/MockNavigator';
import { mockChannel, mockLao, mockReduxAction } from '__tests__/utils';
import FeatureContext from 'core/contexts/FeatureContext';
import { HOME_FEATURE_IDENTIFIER, HomeReactContext } from 'features/home/interface';
import { resubscribeToLao } from 'features/lao/functions';
import { LaoHooks } from 'features/lao/hooks';
import { laoReducer, setCurrentLao } from 'features/lao/reducer';

import ConnectOpenScan from '../ConnectOpenScan';

jest.mock('react-qr-reader');
jest.mock('websocket');

jest.mock('@react-navigation/core', () => {
  const actualNavigation = jest.requireActual('@react-navigation/core');

  const mockNavigate = jest.fn();
  const mockAddListener = jest.fn();

  return {
    ...actualNavigation,
    useNavigation: () => ({
      navigate: mockNavigate,
      addListener: mockAddListener,
    }),
  };
});

// Is mocked
// eslint-disable-next-line react-hooks/rules-of-hooks
// const { navigate: mockNavigate, addListener } = useNavigation();

/*
const didFocus = () =>
  // call focus event listener
  (addListener as jest.Mock).mock.calls
    .filter(([eventName]) => eventName === 'focus')
    .forEach((args) => args[1]());
*/

const mockConnection = 0;

jest.mock('core/network', () => {
  return {
    ...jest.requireActual('core/network'),
    subscribeToChannel: jest.fn(() => Promise.resolve()),
    connect: jest.fn(() => Promise.resolve(mockConnection)),
  };
});

beforeEach(jest.clearAllMocks);

const contextValue = {
  [HOME_FEATURE_IDENTIFIER]: {
    addLaoServerAddress: () => mockReduxAction,
    useCurrentLaoId: LaoHooks.useCurrentLaoId,
    getLaoChannel: () => mockChannel,
    LaoList: () => null,
    connectToTestLao: () => {},
    homeNavigationScreens: [],
    requestCreateLao: () => Promise.resolve(mockChannel),
    useLaoList: () => [],
    useDisconnectFromLao: () => () => {},
    getLaoById: () => mockLao,
    resubscribeToLao,
  } as HomeReactContext,
};

const mockStore = createStore(combineReducers(laoReducer));
mockStore.dispatch(setCurrentLao(mockLao.toState()));

// TODO: Fix react-qr-reader for commented tests to work
describe('ConnectOpenScan', () => {
  it('renders correctly', () => {
    const component = render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={ConnectOpenScan} />
        </FeatureContext.Provider>
      </Provider>,
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  /*
  it('can connect to a lao', async () => {
    render(
      <Provider store={mockStore}>
        <FeatureContext.Provider value={contextValue}>
          <MockNavigator component={ConnectOpenScan} />
        </FeatureContext.Provider>
      </Provider>,
    );

    act(didFocus);

    fakeQrReaderScan(
      new ConnectToLao({
        lao: mockLaoId,
        servers: [mockAddress],
      }).toJson(),
    );

    await waitFor(() => {
      expect(subscribeToChannel).toHaveBeenCalledWith(
        mockLaoIdHash,
        expect.anything(),
        getLaoChannel(mockLaoId),
        expect.anything(),
      );

      expect(mockNavigate).toHaveBeenCalledTimes(1);
    });
  });
  */
});
