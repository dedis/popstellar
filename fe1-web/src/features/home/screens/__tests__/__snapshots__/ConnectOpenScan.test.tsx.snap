// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectOpenScan renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  MockScreen
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            Object {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                            "overflow": "hidden",
                          }
                        }
                      >
                        <View
                          style={
                            Object {
                              "bottom": 0,
                              "left": 0,
                              "position": "absolute",
                              "right": 0,
                              "top": 0,
                              "zIndex": 0,
                            }
                          }
                        />
                        <View
                          style={
                            Object {
                              "flex": 1,
                              "flexDirection": "column",
                              "margin": 16,
                            }
                          }
                        >
                          <View
                            style={
                              Object {
                                "bottom": 0,
                                "flex": 1,
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "flex": 1,
                                  "flexDirection": "row",
                                  "justifyContent": "space-between",
                                }
                              }
                            >
                              <View>
                                <View
                                  style={
                                    Object {
                                      "backgroundColor": "#fff",
                                      "borderRadius": 16,
                                      "padding": 8,
                                    }
                                  }
                                >
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onBlur={[Function]}
                                    onClick={[Function]}
                                    onFocus={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <RNGestureHandlerButton
                                      collapsable={false}
                                      exclusive={true}
                                      handlerTag={2}
                                      handlerType="NativeViewGestureHandler"
                                      onGestureEvent={[Function]}
                                      onGestureHandlerEvent={[Function]}
                                      onGestureHandlerStateChange={[Function]}
                                      onHandlerStateChange={[Function]}
                                      rippleColor={0}
                                      touchSoundDisabled={false}
                                    >
                                      <View
                                        accessible={true}
                                        collapsable={false}
                                        style={
                                          Object {
                                            "opacity": 1,
                                          }
                                        }
                                      >
                                        <View>
                                          {"name":"ios-close","size":25,"color":"#3742fa"}
                                        </View>
                                      </View>
                                    </RNGestureHandlerButton>
                                  </View>
                                </View>
                              </View>
                              <View>
                                <View
                                  style={
                                    Object {
                                      "backgroundColor": "#fff",
                                      "borderRadius": 16,
                                      "padding": 8,
                                    }
                                  }
                                >
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onBlur={[Function]}
                                    onClick={[Function]}
                                    onFocus={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                    testID="launch_selector"
                                  >
                                    <RNGestureHandlerButton
                                      collapsable={false}
                                      exclusive={true}
                                      handlerTag={3}
                                      handlerType="NativeViewGestureHandler"
                                      onGestureEvent={[Function]}
                                      onGestureHandlerEvent={[Function]}
                                      onGestureHandlerStateChange={[Function]}
                                      onHandlerStateChange={[Function]}
                                      rippleColor={0}
                                      touchSoundDisabled={false}
                                    >
                                      <View
                                        accessible={true}
                                        collapsable={false}
                                        style={
                                          Object {
                                            "marginBottom": 8,
                                            "opacity": 1,
                                          }
                                        }
                                      >
                                        <View>
                                          {"name":"ios-create","size":25,"color":"#3742fa"}
                                        </View>
                                      </View>
                                    </RNGestureHandlerButton>
                                  </View>
                                  <View
                                    accessible={true}
                                    collapsable={false}
                                    focusable={true}
                                    onBlur={[Function]}
                                    onClick={[Function]}
                                    onFocus={[Function]}
                                    onResponderGrant={[Function]}
                                    onResponderMove={[Function]}
                                    onResponderRelease={[Function]}
                                    onResponderTerminate={[Function]}
                                    onResponderTerminationRequest={[Function]}
                                    onStartShouldSetResponder={[Function]}
                                  >
                                    <RNGestureHandlerButton
                                      collapsable={false}
                                      exclusive={true}
                                      handlerTag={4}
                                      handlerType="NativeViewGestureHandler"
                                      onGestureEvent={[Function]}
                                      onGestureHandlerEvent={[Function]}
                                      onGestureHandlerStateChange={[Function]}
                                      onHandlerStateChange={[Function]}
                                      rippleColor={0}
                                      touchSoundDisabled={false}
                                    >
                                      <View
                                        accessible={true}
                                        collapsable={false}
                                        style={
                                          Object {
                                            "opacity": 1,
                                          }
                                        }
                                      >
                                        <View>
                                          {"name":"ios-code","size":25,"color":"#3742fa"}
                                        </View>
                                      </View>
                                    </RNGestureHandlerButton>
                                  </View>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
